README.md
# Overview

Simple command-line application which concatenates a list of photos into a single photo
showing them in a matrix (in other words, a grid). 

# Installation

First of all, you need [Python](https://www.python.org/) 3.7 or greater installed in your system.

Then, as `photomatrix` is a command line application, open a terminal and be ready to type commands.

## Install automatically with pip

`photomatrix` is available from the Python Package Index ([PyPI](https://pypi.org/)), so installable
with [pip](https://pip.pypa.io/):

```
pip3 install photomatrix
```

You should now have `photomatrix` available in your shell. For example, print the command help:

```
photomatrix -h
```

## Build locally

Alternatively, you can build locally from source code. First, clone this project from Github.

Cd into the root folder.

Create a virtual environment (e.g. named `my_env`) and activate it:

```
python3 -m venv my_venv
source my_venv/bin/activate
```

Install the dev dependencies and the app dependencies into the virtual environment:

```
pip3 install -r requirements_dev.txt
pip3 install -e .
```

Run like this:

```
python3 -m photomatrix -h
```

# Usage

The basic usage is:

```
photomatrix INPUT_IMAGES OUTPUT_IMAGE
```

where all the images found in the specified input path will be loaded, sorted, combined in a matrix
and the result will be written in the specified output path.  

Further options that you can customize are:
* Number of rows vs columns
* Resize and crop
* Sorting
* Border
* Text to print (e.g. filename or Exif original date)

See the command help (with `-h`) for more info on these options.

# Examples

The following examples use these test input images:

<img src="data/input/01.jpg?raw=true" width="150">

up to

<img src="data/input/12.jpg?raw=true" width="150">

## Matrix with border

To build a matrix with a simple border:

```
photomatrix "data/input/*.jpg" data/output/matrix_border.jpg --border-width-ratio 0.1
```

resulting in:

<img src="data/output/matrix_border.jpg?raw=true" width="600">

## Matrix sorted and with text

To build a matrix with inverse filename order, printing the filename in top center position:

```
photomatrix "data/input/*.jpg" data/output/matrix_sort_text.jpg --sort filename_desc --text-type filename --text-position top_center
```

resulting in:

<img src="data/output/matrix_sort_text.jpg?raw=true" width="600">

# Integration Tests

Run the integration tests with:

```
python3 -m pytest tests/integration
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.5] - 2021-01-09
- Added flag to print version in a single-sourced way

## [0.0.4] - 2020-12-27
- Added integration tests

## [0.0.3] - 2020-05-03
- Use a portable default font and let the user override it if needed

## [0.0.2] - 2020-04-22
- Added console script entry point

## [0.0.1] - 2020-04-19
- First release

tests/integration/conftest.py
from PIL import Image, ImageChops
import pytest
from pathlib import Path

INTEGRATION_TESTS_DIR = Path(__file__).parent


class ImagePaths:
    """Build paths to test images."""

    def __init__(self, tmp_path):
        self.tmp_path = tmp_path

    @staticmethod
    def abs_path(relative_path):
        """Given a path relative to the integration tests folder, return its
        absolute path.
        """
        return INTEGRATION_TESTS_DIR.joinpath(relative_path).resolve()

    @staticmethod
    def input(relative_path):
        """Return the absolute path to an image in the input folder"""
        return ImagePaths.abs_path(Path('input') / relative_path)

    @staticmethod
    def expected(relative_path):
        """Return the absolute path to an image in the expected folder"""
        return ImagePaths.abs_path(Path('expected') / relative_path)

    def actual(self, relative_path):
        """Return the absolute path to an image in the actual folder"""
        return self.tmp_path / 'actual' / relative_path


@pytest.fixture
def img_paths(tmp_path):
    return ImagePaths(tmp_path)


class ImageDiff:
    """Check the difference between test images."""

    @staticmethod
    def assert_equal(actual_img_path, expected_img_path):
        """Check whether the actual image produced by the program is equal to
        a known expected image.
        """
        actual_img = Image.open(actual_img_path).convert('RGB')
        expected_img = Image.open(expected_img_path).convert('RGB')
        diff = ImageChops.difference(actual_img, expected_img)
        if diff.getbbox():
            pytest.fail(f'Image {actual_img_path} is different from expected {expected_img_path}')


@pytest.fixture
def img_diff():
    return ImageDiff()


tests/integration/test_matrix.py
from PIL.ImageColor import colormap
from photomatrix import __main__ as photomatrix


def test_basic(img_paths, img_diff):
    input_imgs = img_paths.input('[ab].png')
    actual_img = img_paths.actual('ab.png')
    expected_img = img_paths.expected('ab.png')

    photomatrix.main(f'{input_imgs} {actual_img}'.split())

    img_diff.assert_equal(actual_img, expected_img)


def test_sort_filename(img_paths, img_diff):
    input_imgs = img_paths.input('[ab].png')
    actual_img = img_paths.actual('ba.png')
    expected_img = img_paths.expected('ba.png')

    photomatrix.main(f'{input_imgs} {actual_img} --sort filename_desc'.split())

    img_diff.assert_equal(actual_img, expected_img)


def test_column_single(img_paths, img_diff):
    input_imgs = img_paths.input('[ab].png')
    actual_img = img_paths.actual('ab_single_column.png')
    expected_img = img_paths.expected('ab_single_column.png')

    photomatrix.main(f'{input_imgs} {actual_img} --columns-num 1'.split())

    img_diff.assert_equal(actual_img, expected_img)


def test_border(img_paths, img_diff):
    input_img = img_paths.input('a.png')
    actual_img = img_paths.actual('a_border.png')
    expected_img = img_paths.expected('a_border.png')

    photomatrix.main(f'{input_img} {actual_img}'
                     f' --border-width-ratio 0.1 --border-color {colormap["blue"]}'.split())

    img_diff.assert_equal(actual_img, expected_img)


def test_resize_half(img_paths, img_diff):
    input_img = img_paths.input('a.png')
    actual_img = img_paths.actual('a_half_size.png')
    expected_img = img_paths.expected('a_half_size.png')

    photomatrix.main(f'{input_img} {actual_img} --resize-ratio 0.5'.split())

    img_diff.assert_equal(actual_img, expected_img)


def test_crop(img_paths, img_diff):
    input_img = img_paths.input('a.png')
    actual_img = img_paths.actual('a_crop.png')
    expected_img = img_paths.expected('a_crop.png')

    photomatrix.main(f'{input_img} {actual_img} --crop-ratio 0.3,0.2,0.7,0.85'.split())

    img_diff.assert_equal(actual_img, expected_img)


def test_text_filename(img_paths, img_diff):
    input_img = img_paths.input('a.png')
    actual_img = img_paths.actual('a_text_filename.png')
    expected_img = img_paths.expected('a_text_filename.png')

    photomatrix.main(
        f'{input_img} {actual_img}'
        ' --text-type filename --text-position top_left'
        f' --text-color {colormap["indigo"]} --text-background-opacity 100 --text-height-ratio 0.3'.split())

    img_diff.assert_equal(actual_img, expected_img)


tests/integration/test_version.py
from photomatrix import __main__ as photomatrix
import pytest
try:
    from importlib import metadata
except ImportError:
    # on Python <3.8 use importlib-metadata package
    import importlib_metadata as metadata


def test_printed_version(capsys):
    expected_version = metadata.version('photomatrix')

    with pytest.raises(SystemExit):
        photomatrix.main(['--version'])

    actual_output = capsys.readouterr().out.rstrip()
    assert actual_output, 'No output found'
    assert actual_output == expected_version


photomatrix/text.py
import os
import photomatrix.exif as exif
from photomatrix.arguments import TextType, TextPosition
from PIL import ImageDraw, ImageFont


def add_text(images, text_config):
    return [add_text_single(img, text_config) for img in images]


def add_text_single(image, text_config):
    text = get_text(image, text_config)
    if text is None:
        return image

    new_image = image.convert(mode='RGB')
    draw = ImageDraw.Draw(new_image, mode='RGBA')
    font = get_font(image, text_config.height_ratio, text_config.font, text)
    rect_xy = get_text_rect(image, font, text, text_config.position)
    draw.rectangle(rect_xy, fill=(255, 255, 255, text_config.background_opacity))
    draw.text(rect_xy[0], text, fill=text_config.color, font=font)
    return new_image


def get_text(image, text_config):
    if text_config.type == TextType.none:
        return None
    else:
        return {
            TextType.filename: lambda img: os.path.basename(img.filename),
            TextType.date_taken: lambda img: exif.get_date_time_original(img, text_config.date_format)
        }[text_config.type](image)


def get_font(image, height_ratio, font_file, text):
    font_size = 0
    font = None
    font_too_small = True
    while font_too_small:
        font = ImageFont.truetype(font_file, font_size)
        font_h = font.getsize(text)[1]
        font_too_small = font_h < height_ratio * image.height
        font_size += 1
    return font


def get_text_rect(image, font, text, text_position):
    img_w, img_h = image.size
    txt_w, txt_h = font.getsize(text)
    margin = img_w * 0.05
    top_left_corner = {
        TextPosition.bottom_center: ((img_w - txt_w)/2, img_h - txt_h - margin),
        TextPosition.bottom_right: (img_w - txt_w - margin, img_h - txt_h - margin),
        TextPosition.bottom_left: (margin, img_h - txt_h - margin),
        TextPosition.top_center: ((img_w - txt_w)/2, margin),
        TextPosition.top_right: (img_w - txt_w - margin, margin),
        TextPosition.top_left: (margin, margin),
        TextPosition.center: ((img_w - txt_w)/2, (img_h - txt_h)/2)
    }[text_position]
    bottom_right_corner = (top_left_corner[0] + txt_w, top_left_corner[1] + txt_h + margin/2)
    return [top_left_corner, bottom_right_corner]


photomatrix/matrix.py
import functools
from math import ceil
from PIL import Image, ImageOps


def build_matrix(images, matrix_config):
    border_width = get_border_width(images, matrix_config.border_width_ratio)
    border_color = matrix_config.border_color
    imgs_with_border = [add_border(img, border_width, border_color) for img in images]
    grid_image = add_border(build_matrix_image(imgs_with_border, matrix_config.columns_num, border_color),
                            border_width, border_color)
    return grid_image


def get_border_width(images, border_width_ratio):
    return ceil(images[0].width * border_width_ratio / 2)


def add_border(image, border_width, border_color):
    if border_width == 0:
        return image
    else:
        return ImageOps.expand(image, border=border_width, fill=border_color)


def build_matrix_image(images, columns_num, color):
    return concat_images(concat_vertically,
                         map(lambda row_imgs: concat_images(concat_horizontally, row_imgs, color),
                             split_in_sublists_of_size(images, columns_num)),
                         color)


def split_in_sublists_of_size(lst, size):
    for i in range(0, len(lst), size):
        yield lst[i: i+size]


def concat_images(concat_function, images, color):
    return functools.reduce(lambda img1, img2: concat_function(img1, img2, color), images)


def concat_horizontally(img1, img2, color):
    new_img = Image.new('RGB', (img1.width + img2.width, max(img1.height, img2.height)), color)
    new_img.paste(img1, (0, 0))
    new_img.paste(img2, (img1.width, 0))
    return new_img


def concat_vertically(img1, img2, color):
    new_img = Image.new('RGB', (max(img1.width, img2.width), img1.height + img2.height), color)
    new_img.paste(img1, (0, 0))
    new_img.paste(img2, (0, img1.height))
    return new_img


photomatrix/preprocess.py
from photomatrix.arguments import Sort
from photomatrix.text import add_text
import math
from operator import attrgetter


def preprocess(images, preprocess_config):
    cropped_images = crop_images(images, preprocess_config.crop_ratio_box)
    resized_images = resize_images(cropped_images, preprocess_config.resize_ratio)
    sorted_images = sort_images(resized_images, preprocess_config.sort)
    return add_text(sorted_images, preprocess_config.text_config)


def crop_images(images, crop_ratio_box):
    return [crop_image(img, crop_ratio_box) for img in images]


def crop_image(image, crop_ratio_box):
    x1, y1, x2, y2 = crop_ratio_box
    w = image.width
    h = image.height
    crop_box = (x1 * w, y1 * h, x2 * w, y2 * h)
    new_image = image.crop(crop_box)
    new_image.filename = image.filename
    return new_image


def resize_images(images, resize_ratio):
    if resize_ratio == 1:
        return images
    return [resize_image(img, resize_ratio) for img in images]


def resize_image(image, resize_ratio):
    new_size = (math.floor(image.width * resize_ratio),
                math.floor(image.height * resize_ratio))
    new_image = image.resize(new_size)
    new_image.filename = image.filename
    return new_image


def sort_images(images, sort):
    (sort_key, is_reversed) = {
        Sort.filename_asc: (attrgetter('filename'), False),
        Sort.filename_desc: (attrgetter('filename'), True)
    }[sort]
    return sorted(images, key=sort_key, reverse=is_reversed)


photomatrix/arguments.py
import argparse
import glob
import os
import re
from enum import Enum
from math import ceil, sqrt
from pathlib import Path
from PIL import Image
from fonts.ttf import SourceSansProSemibold as DefaultFont
try:
    from importlib import metadata
except ImportError:
    # on Python <3.8 use importlib-metadata package
    import importlib_metadata as metadata


class RunConfig:
    def __init__(self, *,
                 input_images,
                 output_image_path,
                 preprocess_config,
                 matrix_config):
        self.input_images = input_images
        self.output_image_path = output_image_path
        self.preprocess_config = preprocess_config
        self.matrix_config = matrix_config


class PreprocessConfig:
    def __init__(self, *, crop_ratio_box, resize_ratio, sort, text_config):
        self.crop_ratio_box = crop_ratio_box
        self.resize_ratio = resize_ratio
        self.sort = sort
        self.text_config = text_config


class MatrixConfig:
    def __init__(self, *, columns_num, border_width_ratio, border_color):
        self.columns_num = columns_num
        self.border_width_ratio = border_width_ratio
        self.border_color = border_color


class BaseEnum(Enum):
    def __str__(self):
        return self.value


class Sort(BaseEnum):
    filename_asc = 'filename_asc'
    filename_desc = 'filename_desc'


class TextConfig:
    def __init__(self, *,
                 text_type,
                 position,
                 color,
                 font,
                 background_opacity,
                 date_format,
                 height_ratio):
        self.type = text_type
        self.position = position
        self.color = color
        self.font = font
        self.background_opacity = background_opacity
        self.date_format = date_format
        self.height_ratio = height_ratio


class TextType(BaseEnum):
    none = 'none'
    filename = 'filename'
    date_taken = 'date_taken'


class TextPosition(BaseEnum):
    bottom_center = 'bottom_center'
    bottom_right = 'bottom_right'
    bottom_left = 'bottom_left'
    top_center = 'top_center'
    top_right = 'top_right'
    top_left = 'top_left'
    center = 'center'


def file_paths(glob_pattern):
    try:
        paths = glob.glob(glob_pattern)
    except Exception as e:
        raise argparse.ArgumentTypeError(f'Failed to parse pattern "{glob_pattern}": {e}')
    else:
        if not paths:
            raise argparse.ArgumentTypeError(f'No files found with pattern "{glob_pattern}"')
        return paths


def load_images(image_paths):
    return list(map(lambda f: Image.open(f), image_paths))


def best_columns_num(images):
    img_num = len(images)
    for col_num in range(ceil(sqrt(img_num)), img_num + 1):
        if img_num % col_num == 0:
            break
    return col_num


def int_range(min_value, max_value):
    def _int_range(string_value):
        err_msg = f'{string_value} is not an int in [{min_value}, {max_value}]'
        try:
            int_value = int(string_value)
        except ValueError:
            raise argparse.ArgumentTypeError(err_msg)
        else:
            if int_value < min_value or int_value > max_value:
                raise argparse.ArgumentTypeError(err_msg)
            return int_value
    return _int_range


def float_range(min_value, max_value):
    def _float_range(string_value):
        err_msg = f'{string_value} is not a decimal number in [{min_value}, {max_value}]'
        try:
            float_value = float(string_value)
        except ValueError:
            raise argparse.ArgumentTypeError(err_msg)
        else:
            if float_value < min_value or float_value > max_value:
                raise argparse.ArgumentTypeError(err_msg)
            return float_value
    return _float_range


def float_positive(string_value):
    err_msg = f'{string_value} is not a positive decimal number'
    try:
        float_value = float(string_value)
    except ValueError:
        raise argparse.ArgumentTypeError(err_msg)
    else:
        if float_value <= 0:
            raise argparse.ArgumentTypeError(err_msg)
        return float_value


def float_non_negative(string_value):
    err_msg = f'{string_value} is not a non-negative decimal number'
    try:
        float_value = float(string_value)
    except ValueError:
        raise argparse.ArgumentTypeError(err_msg)
    else:
        if float_value < 0:
            raise argparse.ArgumentTypeError(err_msg)
        return float_value


def ratio_box(string_value):
    err_msg = f'{string_value} has not the form "x1,y1,x2,y2" where each number is a decimal in [0, 1]'
    try:
        x1, y1, x2, y2 = map(float, string_value.split(','))
        box = (x1, y1, x2, y2)
    except ValueError:
        raise argparse.ArgumentTypeError(err_msg)
    else:
        for i in box:
            if i < 0 or i > 1:
                raise argparse.ArgumentTypeError(err_msg)
    return box


def hexadecimal_color(string_value):
    if not re.match("#[a-f0-9]{6}$", string_value):
        raise argparse.ArgumentTypeError(f'{string_value} is not a #rrggbb hexadecimal color string')
    return string_value


def parse_arguments(args):
    parser = argparse.ArgumentParser(prog='photomatrix', description='Concat photos together in a matrix.')
    parser.add_argument('--version', action='version', version=metadata.version(parser.prog))
    parser.add_argument('input_images', type=file_paths,
                        help='the path to the images to be processed. '
                             'Can contain *, ?, and character ranges expressed with [].')
    parser.add_argument('output_image', type=Path,
                        help='the image resulting from the processing')
    parser.add_argument('--columns-num', type=int,
                        help='the number of columns in the matrix (otherwise a sensible default will be found)')
    parser.add_argument('--sort', type=Sort, choices=list(Sort), default=Sort.filename_asc,
                        help='order of the images in the matrix, starting from top-left and then row by row '
                             f'(defaults to {Sort.filename_asc})')
    parser.add_argument('--border-width-ratio', type=float_non_negative, default=0,
                        help='Width of the border to add to all images in the matrix, '
                             'expressed as a ratio of the width of the first image of the matrix. '
                             'Must be a decimal number >= 0. Defaults to 0, meaning no border.')
    parser.add_argument('--border-color', type=hexadecimal_color, default="#000000",
                        help='The border color expressed as #rrggbb hexadecimal color string. '
                             'Defaults to black.')
    parser.add_argument('--resize-ratio', type=float_positive, default=1,
                        help='Resize ratio to apply to each image. Must be a positive decimal number. '
                             'For example, 0.2 means that the images are reduced to 20%% of their original size. '
                             'Defaults to 1, i.e. keep the original size.')
    parser.add_argument('--crop-ratio', type=ratio_box, default="0,0,1,1",
                        help='Crop to apply to each image. '
                             'The crop is expressed as a box in the form "x1,y1,x2,y2" where x1,y1 is the upper-left '
                             'corner of the box and x2,y2 is the lower-right corner. '
                             'Coordinates are not absolute pixels, instead they are decimal ratios '
                             'over the image width (for x1 and x2) or height (for y1 and y2). '
                             'Defaults to "0,0,1,1", i.e. keep the original image without cropping.')
    parser.add_argument('--text-type', type=TextType, choices=list(TextType), default=TextType.none,
                        help='text to be added to each image '
                             f'(defaults to {TextType.none})')
    parser.add_argument('--text-position', type=TextPosition, choices=list(TextPosition),
                        default=TextPosition.bottom_center,
                        help='the position of the text relative to the image '
                             f'(defaults to {TextPosition.bottom_center})')
    parser.add_argument('--text-color', type=hexadecimal_color, default="#000000",
                        help='The text color expressed as #rrggbb hexadecimal color string. '
                             'Defaults to black.')
    parser.add_argument('--text-font', default=DefaultFont,
                        help='The text font. Must be a TrueType or OpenType font file, either '
                             'an absolute path or only the filename if located in a system default folder. '
                             f'Defaults to {os.path.basename(DefaultFont)}')
    parser.add_argument('--text-background-opacity', type=int_range(0, 255), default=0,
                        help='Opacity of the text background, use to make the text more distinguishable. '
                             'Must be between 0 and 255. Defaults to 0, meaning no text background.')
    parser.add_argument('--text-date-format', default='%d/%m/%Y',
                        help=f'format of the date in case of {TextType.date_taken}, see possible codes: '
                             'https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes')
    parser.add_argument('--text-height-ratio', type=float_range(0, 1), default=0.2,
                        help='Ratio of the image height over the text height. '
                             'Must be a decimal number between 0 and 1. '
                             'For example a value of 0.1 means that the text height is 10%% of the image height. '
                             'Defaults to 0.2.')
    return parser.parse_args(args)


def parse_run_config(args):
    args = parse_arguments(args)

    input_images = load_images(args.input_images)
    columns_num = args.columns_num if args.columns_num is not None else best_columns_num(input_images)
    text_config = TextConfig(text_type=args.text_type,
                             position=args.text_position,
                             color=args.text_color,
                             font=args.text_font,
                             background_opacity=args.text_background_opacity,
                             date_format=args.text_date_format,
                             height_ratio=args.text_height_ratio)
    preprocess_config = PreprocessConfig(crop_ratio_box=args.crop_ratio,
                                         resize_ratio=args.resize_ratio,
                                         sort=args.sort,
                                         text_config=text_config)
    matrix_config = MatrixConfig(columns_num=columns_num,
                                 border_width_ratio=args.border_width_ratio,
                                 border_color=args.border_color)

    return RunConfig(input_images=input_images,
                     output_image_path=args.output_image,
                     preprocess_config=preprocess_config,
                     matrix_config=matrix_config)
